[Homework Questions Link](https://github.com/DataTalksClub/stock-markets-analytics-zoomcamp/blob/main/cohorts/2026/homework2.md)

## Question 1: [IPO] Withdrawn IPOs by Company Type

### What is the total withdrawn IPO value (in $ millions) for the company class with the highest total withdrawal value?

In [ ]:
import pandas as pd
import re
import numpy as np

In [ ]:
url = "https://www.iposcoop.com/ipos-recently-filed/"
ipos_recently_filed = pd.read_html(url)[0]

# look at the ipo data
ipos_recently_filed.head(3)

,File Date,Company,Symbol,Managers,Shares (millions),Price Low,Price High,Est $ Vol (millions),Expected To Trade,SCOOP Rating
0,2026-09-28,Encore Medical Inc.,EMI,The Oak Ridge Financial Services Group/Network...,3.0,$5.00,$5.00,$15.00,Wednesday,S/O
1,2026-09-25,Elevation Acquisition Group,ELEVU,Maxim Group,10.0,$10.00,$10.00,$100.00,TBA,S/O
2,2026-09-25,Holtec Nuclear (Withdrawn),HNUC,J.P.Morgan/Guggenheim Securities/Goldman Sachs...,50.0,$15.00,$18.00,$825.00,Withdrawn,S/O


In [ ]:
ipos_recently_filed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   File Date             500 non-null    object 
 1   Company               500 non-null    object 
 2   Symbol                500 non-null    object 
 3   Managers              500 non-null    object 
 4   Shares (millions)     500 non-null    float64
 5   Price Low             464 non-null    object 
 6   Price High            464 non-null    object 
 7   Est $ Vol (millions)  495 non-null    object 
 8   Expected To Trade     500 non-null    object 
 9   SCOOP Rating          500 non-null    object 
dtypes: float64(1), object(9)
memory usage: 39.2+ KB


In [ ]:
# filter to 'Withdrawn' only for Expected to Trade
withdrawn_df = ipos_recently_filed[ipos_recently_filed['Expected To Trade']=='Withdrawn']
display(withdrawn_df.shape)

(35, 10)

In [ ]:
# classifies company type based on keywords
def classify_company_type(company_name):
    company_name_lower = str(company_name).lower()
    if 'technologies' in company_name_lower:
        return 'Technologies'
    elif 'acquisition corp' in company_name_lower or \
         'acquisition corporation' in company_name_lower or \
         re.search(r'\bcorp\b', company_name_lower):
        return 'Acquisition Corp'
    elif 'inc' in company_name_lower or 'incorporated' in company_name_lower:
        return 'Inc.'
    elif 'group' in company_name_lower:
        return 'Group'
    elif 'ltd' in company_name_lower or 'limited' in company_name_lower:
        return 'Limited'
    elif 'holdings' in company_name_lower or 'holding' in company_name_lower:
        return 'Holdings'
    else:
        return 'Other'

withdrawn_df['Company Type'] = withdrawn_df['Company'].apply(classify_company_type)
withdrawn_df.head(3)

/tmp/ipykernel_1996/1360166711.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  withdrawn_df['Company Type'] = withdrawn_df['Company'].apply(classify_company_type)


,File Date,Company,Symbol,Managers,Shares (millions),Price Low,Price High,Est $ Vol (millions),Expected To Trade,SCOOP Rating,Company Type
2,2026-09-25,Holtec Nuclear (Withdrawn),HNUC,J.P.Morgan/Guggenheim Securities/Goldman Sachs...,50.00,$15.00,$18.00,$825.00,Withdrawn,S/O,Other
24,2026-09-18,Rise Smart Group (Withdrawn),RSHL,American Trust Investment Services/Prime Numbe...,1.80,$4.00,$4.00,$7.20,Withdrawn,S/O,Group
27,2026-09-17,C2 Capital Group (Withdrawn),CCLV,Dominari Securities,3.75,$4.00,$5.00,$16.88,Withdrawn,S/O,Group


In [ ]:
def find_avg_price(row: pd.DataFrame):
  low = pd.to_numeric(str(row["Price Low"]).replace('$', ''), errors="coerce")
  high = pd.to_numeric(str(row["Price High"]).replace('$', ''), errors="coerce")

  return (low + high) / 2

withdrawn_df['Avg_price'] = withdrawn_df.apply(find_avg_price, axis=1)
withdrawn_df.head(3)

/tmp/ipykernel_1996/597125063.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  withdrawn_df['Avg_price'] = withdrawn_df.apply(find_avg_price, axis=1)


,File Date,Company,Symbol,Managers,Shares (millions),Price Low,Price High,Est $ Vol (millions),Expected To Trade,SCOOP Rating,Company Type,Avg_price
2,2026-09-25,Holtec Nuclear (Withdrawn),HNUC,J.P.Morgan/Guggenheim Securities/Goldman Sachs...,50.00,$15.00,$18.00,$825.00,Withdrawn,S/O,Other,16.5
24,2026-09-18,Rise Smart Group (Withdrawn),RSHL,American Trust Investment Services/Prime Numbe...,1.80,$4.00,$4.00,$7.20,Withdrawn,S/O,Group,4.0
27,2026-09-17,C2 Capital Group (Withdrawn),CCLV,Dominari Securities,3.75,$4.00,$5.00,$16.88,Withdrawn,S/O,Group,4.5


In [ ]:
withdrawn_df["Shares (millions)"] = pd.to_numeric(withdrawn_df['Shares (millions)'], errors="coerce")
withdrawn_df['Est $ Vol (millions)'] = pd.to_numeric(withdrawn_df['Est $ Vol (millions)'].str.replace('$', '').replace(',', ''), errors="coerce")
withdrawn_df.head(3)

/tmp/ipykernel_1996/761777133.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  withdrawn_df["Shares (millions)"] = pd.to_numeric(withdrawn_df['Shares (millions)'], errors="coerce")
/tmp/ipykernel_1996/761777133.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  withdrawn_df['Est $ Vol (millions)'] = pd.to_numeric(withdrawn_df['Est $ Vol (millions)'].str.replace('$', '').replace(',', ''), errors="coerce")


,File Date,Company,Symbol,Managers,Shares (millions),Price Low,Price High,Est $ Vol (millions),Expected To Trade,SCOOP Rating,Company Type,Avg_price
2,2026-09-25,Holtec Nuclear (Withdrawn),HNUC,J.P.Morgan/Guggenheim Securities/Goldman Sachs...,50.00,$15.00,$18.00,825.00,Withdrawn,S/O,Other,16.5
24,2026-09-18,Rise Smart Group (Withdrawn),RSHL,American Trust Investment Services/Prime Numbe...,1.80,$4.00,$4.00,7.20,Withdrawn,S/O,Group,4.0
27,2026-09-17,C2 Capital Group (Withdrawn),CCLV,Dominari Securities,3.75,$4.00,$5.00,16.88,Withdrawn,S/O,Group,4.5


In [ ]:
withdrawn_df['Shares_offered_value'] = np.where(
    pd.isna(withdrawn_df['Shares (millions)'] * withdrawn_df['Avg_price']),
    withdrawn_df['Est $ Vol (millions)'],
    withdrawn_df['Shares (millions)'] * withdrawn_df['Avg_price']
)
withdrawn_df.head()

/tmp/ipykernel_1996/804680285.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  withdrawn_df['Shares_offered_value'] = np.where(


,File Date,Company,Symbol,Managers,Shares (millions),Price Low,Price High,Est $ Vol (millions),Expected To Trade,SCOOP Rating,Company Type,Avg_price,Shares_offered_value
2,2026-09-25,Holtec Nuclear (Withdrawn),HNUC,J.P.Morgan/Guggenheim Securities/Goldman Sachs...,50.00,$15.00,$18.00,825.00,Withdrawn,S/O,Other,16.5,825.000
24,2026-09-18,Rise Smart Group (Withdrawn),RSHL,American Trust Investment Services/Prime Numbe...,1.80,$4.00,$4.00,7.20,Withdrawn,S/O,Group,4.0,7.200
27,2026-09-17,C2 Capital Group (Withdrawn),CCLV,Dominari Securities,3.75,$4.00,$5.00,16.88,Withdrawn,S/O,Group,4.5,16.875
39,2026-09-15,OTSAW Ltd. (Withdrawn),OTSA,Aegis Capital,4.40,$4.50,$5.50,22.00,Withdrawn,S/O,Limited,5.0,22.000
45,2026-09-10,Motive Technologies (Withdrawn),MTVE,J.P.Morgan/Citigroup/Barclays/Jefferies/RBC Ca...,0.00,NaN,NaN,100.00,Withdrawn,S/O,Technologies,NaN,100.000


In [ ]:
agg_df = withdrawn_df.groupby(['Company Type']) \
.Shares_offered_value.agg(['sum']) \
.rename({'sum': 'sum_shares_offered_value'}, axis=1)
agg_df

,sum_shares_offered_value
Company Type,
Acquisition Corp,499.9850
Group,56.5750
Holdings,311.6575
Inc.,351.0000
Limited,219.2500
Other,1115.4450
Technologies,184.9000


In [ ]:
agg_df.sort_values(by='sum_shares_offered_value', ascending=False)

,sum_shares_offered_value
Company Type,
Other,1115.4450
Acquisition Corp,499.9850
Inc.,351.0000
Holdings,311.6575
Limited,219.2500
Technologies,184.9000
Group,56.5750


### Answer: The class 'Acquisition Corp' has the highest total value of withdrawals (outside of Other) with $499.9850.

## Question 2: [IPO] Median Sharpe Ratio for 2025 IPOs (First 8 Months)

### What is the median Sharpe ratio (as of 11 September 2026) for companies that went public before 1 September 2025?

In [58]:
import pandas as pd
from time import sleep
import yfinance as yf
import numpy as np
import datetime as dt

In [59]:
url = "https://www.iposcoop.com/2025-pricings/"
ipos25 = pd.read_html(url)[0]
display(ipos25.head(3))
display(ipos25.info())

# convert datatypes to correct format
ipos25["Offer Date"] = pd.to_datetime(ipos25["Offer Date"])
ipos25["Return"] = pd.to_numeric(ipos25["Return"].str.replace("%", ''), errors="coerce")

,Company,Symbol,Industry,Offer Date,Shares (millions),Offer Price,1st Day Close,Current Price,Return,SCOOP Rating
0,Vine Hill Capital Investment Corp. II,VHCPU,Blank Check,12/18/2025,20.0,$10.00,$0.00,$0.00,0.00%,S/O
1,Andersen Group,ANDG,Consumer Services,12/17/2025,11.0,$16.00,$23.50,$23.79,48.69%,S/O
2,Medline Inc.,MDLN,Health Care,12/17/2025,216.0,$29.00,$41.00,$43.83,51.14%,S/O


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 231 entries, 0 to 230
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Company            231 non-null    object 
 1   Symbol             231 non-null    object 
 2   Industry           231 non-null    object 
 3   Offer Date         231 non-null    object 
 4   Shares (millions)  231 non-null    float64
 5   Offer Price        231 non-null    object 
 6   1st Day Close      231 non-null    object 
 7   Current Price      231 non-null    object 
 8   Return             231 non-null    object 
 9   SCOOP Rating       231 non-null    object 
dtypes: float64(1), object(9)
memory usage: 18.2+ KB


None

In [60]:
# filter to only before sep 1 2025 offer date and remove those with 0% return
filtered_df = ipos25[(ipos25["Offer Date"] < "2025-09-01") & (ipos25["Return"] != 0)]
# filtered_df = ipos25[ipos25["Offer Date"] < "2025-09-01"]
# filtered_df = filtered_df[filtered_df["Return"] != 0.0]
display(filtered_df.head(3))
display(filtered_df.shape)

,Company,Symbol,Industry,Offer Date,Shares (millions),Offer Price,1st Day Close,Current Price,Return,SCOOP Rating
68,"Picard Medical, Inc.",PMI,Health Care,2025-08-29,4.3,$4.00,$4.59,$1.68,-58.00,S/O
69,GrowHub Limited (The),TGHL,Technology,2025-08-28,3.7,$4.00,$3.38,$0.40,-90.00,S/O
70,TryHard Holdings Limited,THH,Consumer Services,2025-08-28,1.5,$4.00,$6.02,$29.47,636.75,S/O


(146, 10)

In [61]:
# get daily stocks prices
tickers = filtered_df['Symbol'].unique()
stocks_df = pd.DataFrame({'A': []})
start = dt.date(year=2024, month=1, day=1)
end = dt.date(year=2026, month=9, day=15)
risk_free_rate_of_return = 0.05
delisted = []

for i, t in enumerate(tickers):
  # print(i, t)
  try:
    # get data for the current ticker
    temp = yf.Ticker(t).history(start=start, end=end, interval="1d")

    # transformations
    temp['ticker'] = t
    temp['Date'] = pd.to_datetime(temp.index.date)
    temp['year'] = temp.index.year
    temp['month'] = temp.index.month
    temp['weekday'] = temp.index.weekday
    temp['close_252d'] = temp.Close.shift(252)
    # 1 year of trading day's growth
    temp['growth_252d'] = temp.Close / temp['close_252d']
    temp['growth_252d_percent'] = (temp.Close / temp['close_252d']) - 1
    # annualized volatility
    temp['volatility'] = temp.Close.rolling(30).std() * np.sqrt(252)
    # sharpe ratio - risk-adjusted performance of an investment (higher is better)
    temp['sharpe'] = (
        temp['growth_252d'] - risk_free_rate_of_return
        ) / temp['volatility']

    # calculate future growth for question 3
    for month_num in range(1, 13, 1):
      temp["future_growth_"+str(month_num)+"_m"] = (
          temp.Close.shift(-21 * month_num) / temp.Close
      )


  except Exception as e:
    print(f"Skipped {t} due to {e}")
    delisted.append(t)
    continue

  else:
    # save into main dataframe
    if stocks_df.empty:
      stocks_df = temp
    else:
      stocks_df = pd.concat([stocks_df, temp], ignore_index=True)

  sleep(1)

ERROR:yfinance:HTTP Error 404: {"quoteSummary":{"result":null,"error":{"code":"Not Found","description":"Quote not found for symbol: MJID"}}}
ERROR:yfinance:$MJID: possibly delisted; no timezone found


Skipped MJID due to 'Index' object has no attribute 'date'


ERROR:yfinance:HTTP Error 404: {"quoteSummary":{"result":null,"error":{"code":"Not Found","description":"Quote not found for symbol: EMPG"}}}
ERROR:yfinance:$EMPG: possibly delisted; no timezone found


Skipped EMPG due to 'Index' object has no attribute 'date'


ERROR:yfinance:$CAEP: possibly delisted; no timezone found


Skipped CAEP due to 'Index' object has no attribute 'date'


ERROR:yfinance:$PTNM: possibly delisted; no timezone found


Skipped PTNM due to 'Index' object has no attribute 'date'


ERROR:yfinance:$AHL: possibly delisted; no timezone found
ERROR:yfinance:$CEPT: possibly delisted; no timezone found


Skipped AHL due to 'Index' object has no attribute 'date'
Skipped CEPT due to 'Index' object has no attribute 'date'


ERROR:yfinance:$SDM: possibly delisted; no timezone found


Skipped SDM due to 'Index' object has no attribute 'date'


ERROR:yfinance:$TBH: possibly delisted; no timezone found


Skipped TBH due to 'Index' object has no attribute 'date'


ERROR:yfinance:$WGRX: possibly delisted; no timezone found


Skipped WGRX due to 'Index' object has no attribute 'date'


ERROR:yfinance:$AGH: possibly delisted; no timezone found


Skipped AGH due to 'Index' object has no attribute 'date'


ERROR:yfinance:$EPWK: possibly delisted; no timezone found


Skipped EPWK due to 'Index' object has no attribute 'date'


ERROR:yfinance:$MTSR: possibly delisted; no timezone found


Skipped MTSR due to 'Index' object has no attribute 'date'


ERROR:yfinance:$SKBL: possibly delisted; no timezone found
ERROR:yfinance:$MCTR: possibly delisted; no timezone found


Skipped SKBL due to 'Index' object has no attribute 'date'
Skipped MCTR due to 'Index' object has no attribute 'date'


In [62]:
stocks_df.tail(3)

,Open,High,Low,Close,Volume,Dividends,Stock Splits,ticker,Date,year,...,future_growth_4_m,future_growth_5_m,future_growth_6_m,future_growth_7_m,future_growth_8_m,future_growth_9_m,future_growth_10_m,future_growth_11_m,future_growth_12_m,Capital Gains
46210,3.50,3.530,3.267,3.35,27400,0.0,0.0,INLF,2026-09-10,2026,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
46211,3.30,3.374,3.210,3.28,37400,0.0,0.0,INLF,2026-09-11,2026,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
46212,3.15,3.280,2.935,2.96,58900,0.0,0.0,INLF,2026-09-14,2026,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [63]:
stocks_filtered = stocks_df[stocks_df.Date == '2026-09-11']
display(stocks_filtered.sharpe.describe())

,sharpe
count,130.000000
mean,inf
std,NaN
min,-0.040147
25%,0.012042
50%,0.050142
75%,0.129390
max,inf


### Answer: The median Sharpe ratio is 0.050142 or 0.04.

## Question 3: [IPO] 'Fixed Months Holding Strategy'

### What is the optimal number of months (1 to 12) to hold a newly IPO'd stock in order to maximize the median growth value?

In [64]:
display(stocks_df.head(3))
display(stocks_df.info())

,Open,High,Low,Close,Volume,Dividends,Stock Splits,ticker,Date,year,...,future_growth_4_m,future_growth_5_m,future_growth_6_m,future_growth_7_m,future_growth_8_m,future_growth_9_m,future_growth_10_m,future_growth_11_m,future_growth_12_m,Capital Gains
0,210.5,356.0,209.5,229.5,126794,0.0,0.0,PMI,2025-08-29,2025,...,0.381264,0.381264,0.209150,0.228758,0.095643,0.058824,0.036819,0.022353,0.016819,NaN
1,242.5,267.5,232.5,257.0,98358,0.0,0.0,PMI,2025-09-02,2025,...,0.340467,0.402724,0.221790,0.185019,0.068482,0.059922,0.034047,0.016732,0.017821,NaN
2,257.0,270.0,250.5,264.5,35960,0.0,0.0,PMI,2025-09-03,2025,...,0.344045,0.362949,0.202268,0.181474,0.039698,0.060491,0.031758,0.013724,0.016975,NaN


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46213 entries, 0 to 46212
Data columns (total 30 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Open                 46213 non-null  float64       
 1   High                 46213 non-null  float64       
 2   Low                  46213 non-null  float64       
 3   Close                46213 non-null  float64       
 4   Volume               46213 non-null  int64         
 5   Dividends            46213 non-null  float64       
 6   Stock Splits         46213 non-null  float64       
 7   ticker               46213 non-null  object        
 8   Date                 46213 non-null  datetime64[ns]
 9   year                 46213 non-null  int32         
 10  month                46213 non-null  int32         
 11  weekday              46213 non-null  int32         
 12  close_252d           13178 non-null  float64       
 13  growth_252d          13178 non-

None

In [65]:
display(filtered_df.tail(3))
display(filtered_df.info())

,Company,Symbol,Industry,Offer Date,Shares (millions),Offer Price,1st Day Close,Current Price,Return,SCOOP Rating
228,"Cantor Equity Partners I, Inc. (Stock-Only SPAC)",CEPO,Blank Check,2025-01-07,20.0,$10.00,$10.05,$10.10,1.0,S/O
229,Zhengye Biotechnology Holding Limited,ZYBT,Health Care,2025-01-07,1.5,$4.00,$4.86,$4.22,5.5,S/O
230,INLIF Ltd.,INLF,Industrials,2025-01-02,2.0,$4.00,$4.45,$1.96,-51.0,S/O


<class 'pandas.core.frame.DataFrame'>
Index: 146 entries, 68 to 230
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Company            146 non-null    object        
 1   Symbol             146 non-null    object        
 2   Industry           146 non-null    object        
 3   Offer Date         146 non-null    datetime64[ns]
 4   Shares (millions)  146 non-null    float64       
 5   Offer Price        146 non-null    object        
 6   1st Day Close      146 non-null    object        
 7   Current Price      146 non-null    object        
 8   Return             146 non-null    float64       
 9   SCOOP Rating       146 non-null    object        
dtypes: datetime64[ns](1), float64(2), object(7)
memory usage: 12.5+ KB


None

In [66]:
# convert 1st day close to numeric
filtered_df["close_1st_day"] = pd.to_numeric(filtered_df["1st Day Close"].str.replace("$", ""), errors="coerce")
entry_df = filtered_df[["Symbol", "Offer Date", "close_1st_day"]]
entry_df.rename(columns={"Symbol": "ticker", "Offer Date": "min_date"}, inplace=True)
entry_df["ticker-mindate"] = entry_df["ticker"] + entry_df["min_date"].dt.strftime("%Y-%m-%d")
entry_df.head(3)

/tmp/ipykernel_920/3220531968.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df["close_1st_day"] = pd.to_numeric(filtered_df["1st Day Close"].str.replace("$", ""), errors="coerce")
/tmp/ipykernel_920/3220531968.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  entry_df.rename(columns={"Symbol": "ticker", "Offer Date": "min_date"}, inplace=True)
/tmp/ipykernel_920/3220531968.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in th

,ticker,min_date,close_1st_day,ticker-mindate
68,PMI,2025-08-29,4.59,PMI2025-08-29
69,TGHL,2025-08-28,3.38,TGHL2025-08-28
70,THH,2025-08-28,6.02,THH2025-08-28


In [67]:
# create join column for later
stocks_df["ticker-date"] = stocks_df["ticker"] + stocks_df["Date"].dt.strftime("%Y-%m-%d")
stocks_df.head(3)

,Open,High,Low,Close,Volume,Dividends,Stock Splits,ticker,Date,year,...,future_growth_5_m,future_growth_6_m,future_growth_7_m,future_growth_8_m,future_growth_9_m,future_growth_10_m,future_growth_11_m,future_growth_12_m,Capital Gains,ticker-date
0,210.5,356.0,209.5,229.5,126794,0.0,0.0,PMI,2025-08-29,2025,...,0.381264,0.209150,0.228758,0.095643,0.058824,0.036819,0.022353,0.016819,NaN,PMI2025-08-29
1,242.5,267.5,232.5,257.0,98358,0.0,0.0,PMI,2025-09-02,2025,...,0.402724,0.221790,0.185019,0.068482,0.059922,0.034047,0.016732,0.017821,NaN,PMI2025-09-02
2,257.0,270.0,250.5,264.5,35960,0.0,0.0,PMI,2025-09-03,2025,...,0.362949,0.202268,0.181474,0.039698,0.060491,0.031758,0.013724,0.016975,NaN,PMI2025-09-03


In [80]:
merged_df = pd.merge(
    entry_df,
    stocks_df,
    how="inner",
    left_on="ticker-mindate",
    right_on="ticker-date",
    validate="many_to_one"
)[["ticker_x", "min_date", "Close", "close_1st_day",
   "future_growth_1_m", "future_growth_2_m", "future_growth_3_m", "future_growth_4_m",
   "future_growth_5_m", "future_growth_6_m", "future_growth_7_m", "future_growth_8_m",
   "future_growth_9_m", "future_growth_10_m", "future_growth_11_m", "future_growth_12_m"
   ]]
display(merged_df.head())

,ticker_x,min_date,Close,close_1st_day,future_growth_1_m,future_growth_2_m,future_growth_3_m,future_growth_4_m,future_growth_5_m,future_growth_6_m,future_growth_7_m,future_growth_8_m,future_growth_9_m,future_growth_10_m,future_growth_11_m,future_growth_12_m
0,PMI,2025-08-29,229.500000,4.59,1.851852,0.714597,0.464052,0.381264,0.381264,0.209150,0.228758,0.095643,0.058824,0.036819,0.022353,0.016819
1,TGHL,2025-08-28,3.380000,3.38,0.491124,0.235503,0.157988,0.108580,0.195266,0.119231,0.082840,0.108284,0.399408,0.166568,0.334320,0.225148
2,THH,2025-08-28,60.200001,6.02,1.172757,1.531561,2.076412,3.719269,0.125083,0.083056,0.056146,0.062126,0.063123,0.048671,0.039701,0.035382
3,CURX,2025-08-26,80.400002,4.02,0.168905,0.164428,0.115920,0.099502,0.087562,0.114428,0.119403,0.106468,0.085323,0.066169,0.048259,0.058209
4,CEPF,2025-08-21,10.120000,10.12,1.004941,1.038538,1.023715,1.017787,1.032609,1.031621,1.010870,1.013834,1.014822,1.023715,1.037549,1.026186


In [109]:
merged_df.describe()

,min_date,Close,close_1st_day,future_growth_1_m,future_growth_2_m,future_growth_3_m,future_growth_4_m,future_growth_5_m,future_growth_6_m,future_growth_7_m,future_growth_8_m,future_growth_9_m,future_growth_10_m,future_growth_11_m,future_growth_12_m
count,128,128.000000,128.000000,128.000000,128.000000,128.000000,128.000000,128.000000,128.000000,128.000000,128.000000,128.000000,128.000000,128.000000,127.000000
mean,2025-05-07 19:07:30,199.071456,13.535859,1.058371,1.187848,1.162128,1.053333,0.982044,1.025638,1.065600,0.953583,1.119225,1.148653,1.023984,0.922933
min,2025-01-02 00:00:00,2.820000,0.000000,0.072941,0.106197,0.090147,0.063927,0.052968,0.042922,0.039041,0.033105,0.026234,0.012262,0.011048,0.004833
25%,2025-03-02 18:00:00,4.807500,3.887500,0.710433,0.575402,0.439757,0.350851,0.342989,0.273588,0.232492,0.188171,0.170216,0.168216,0.143868,0.135503
50%,2025-05-08 12:00:00,18.468416,4.915000,0.922104,0.882813,0.760258,0.718638,0.655290,0.670433,0.589150,0.548434,0.551352,0.468903,0.415262,0.438788
75%,2025-07-23 00:00:00,38.897499,18.657500,1.098038,1.151279,1.149382,1.200740,1.069129,1.036421,1.017494,1.021852,1.021540,1.025467,1.065346,1.040961
max,2025-08-29 00:00:00,14240.000000,115.500000,8.787346,16.066783,16.344464,11.386643,8.182777,9.894552,20.741478,21.370826,27.627298,34.687666,19.595802,22.044619
std,NaN,1313.489157,17.283234,1.032996,1.751916,1.763186,1.369993,1.237204,1.496502,2.175021,2.022384,2.985701,3.425724,2.123226,2.127114


### Answer: The optimal holding period is only 1 month with median future growth of 0.922104. The value of the newly IPO'd stocks tend to decrease in its first 12 months after the IPO.

## Question 4: [Strategy] Simple RSI-Based Trading Strategy

### What is the total profit (in $ thousands) you would have earned by investing $1000 every time a stock was oversold (RSI < 30)?

In [2]:
import gdown
import pandas as pd
file_id = "1grCTCzMZKY5sJRtdbLVCXg8JXA8VPyg-"
gdown.download(f"https://drive.google.com/uc?id={file_id}", "data.parquet", quiet=False)
df = pd.read_parquet("data.parquet", engine="pyarrow")

Downloading...
From (original): https://drive.google.com/uc?id=1grCTCzMZKY5sJRtdbLVCXg8JXA8VPyg-
From (redirected): https://drive.google.com/uc?id=1grCTCzMZKY5sJRtdbLVCXg8JXA8VPyg-&confirm=t&uuid=0f12f307-3816-43e1-81bd-5f615a8ce54a
To: /content/data.parquet
100%|██████████| 130M/130M [00:04<00:00, 29.0MB/s]


,Open,High,Low,Close_x,Volume,Dividends,Stock Splits,Ticker,Year,Month,...,growth_brent_oil_7d,growth_brent_oil_30d,growth_brent_oil_90d,growth_brent_oil_365d,growth_btc_usd_1d,growth_btc_usd_3d,growth_btc_usd_7d,growth_btc_usd_30d,growth_btc_usd_90d,growth_btc_usd_365d
0,0.054277,0.062259,0.054277,0.059598,1.031789e+09,0.0,0.0,MSFT,1986,1986-03-01,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,0.059598,0.062791,0.059598,0.061726,3.081600e+08,0.0,0.0,MSFT,1986,1986-03-01,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,0.061726,0.063323,0.061726,0.062791,1.331712e+08,0.0,0.0,MSFT,1986,1986-03-01,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [3]:
display(df.head(3))
display(df.info())

,Open,High,Low,Close_x,Volume,Dividends,Stock Splits,Ticker,Year,Month,...,growth_brent_oil_7d,growth_brent_oil_30d,growth_brent_oil_90d,growth_brent_oil_365d,growth_btc_usd_1d,growth_btc_usd_3d,growth_btc_usd_7d,growth_btc_usd_30d,growth_btc_usd_90d,growth_btc_usd_365d
0,0.054277,0.062259,0.054277,0.059598,1.031789e+09,0.0,0.0,MSFT,1986,1986-03-01,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,0.059598,0.062791,0.059598,0.061726,3.081600e+08,0.0,0.0,MSFT,1986,1986-03-01,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,0.061726,0.063323,0.061726,0.062791,1.331712e+08,0.0,0.0,MSFT,1986,1986-03-01,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


<class 'pandas.core.frame.DataFrame'>
Index: 229932 entries, 0 to 5690
Columns: 203 entries, Open to growth_btc_usd_365d
dtypes: datetime64[ns](3), float64(129), int32(64), int64(5), object(2)
memory usage: 301.7+ MB


None

In [56]:
oversold_signal = 30

# filter date range
filtered = df[(df["Date"] > "2000-01-01") & (df["Date"] < "2025-06-01")]
# filter on buy signal
filtered = filtered[filtered["rsi"] < oversold_signal]
# trim down columns
filtered = filtered[[
    "Ticker", "Date", "Close_x", "rsi", "growth_future_30d"
]]
# calculate income
filtered["income"] = 1000 * (filtered["growth_future_30d"] - 1)
filtered["profitable"] = filtered["income"] > 0
display(filtered.shape)
display(filtered.head())
display((filtered["profitable"] == True).sum() / filtered.shape[0])

(5206, 7)

,Ticker,Date,Close_x,rsi,growth_future_30d,income,profitable
3562,MSFT,2000-04-14,22.719513,28.599774,0.898820,-101.180298,False
3567,MSFT,2000-04-24,20.420746,28.183709,1.102251,102.251355,True
3667,MSFT,2000-09-14,20.171711,28.744488,0.975309,-24.691404,False
3668,MSFT,2000-09-15,19.673643,24.777933,0.985394,-14.605724,False
3669,MSFT,2000-09-18,19.309669,22.350677,0.972222,-27.778030,False


np.float64(0.5512869765655013)

In [52]:
filtered.describe()

,Date,Close_x,rsi,growth_future_30d,income
count,5206,5206.000000,5206.000000,5206.000000,5206.000000
mean,2012-04-01 01:40:57.625816576,187.375770,25.680893,1.012640,12.640336
min,2000-01-27 00:00:00,0.056319,0.000000,0.553807,-446.193002
25%,2004-07-28 00:00:00,14.639974,24.277473,0.980753,-19.247300
50%,2011-08-18 00:00:00,34.139362,26.649822,1.006329,6.328955
75%,2019-08-06 00:00:00,158.819199,28.580233,1.041476,41.476114
max,2025-04-22 00:00:00,4074.156982,29.999511,3.018886,2018.886052
std,NaN,449.939730,4.345780,0.094235,94.234572


In [51]:
net_income = filtered["income"].sum()
display(net_income)

np.float64(65805.58960747934)

### Answer: The net income is about $65,805.

## Q5. [Exploratory, Optional] Predicting a Positive-Return IPO

Most of the strategies for investing in IPOs deliver negative average and median returns (and even the 75th percentile).

### Question: How would you change the strategy if you want to increase profitability?

### Answer: I would be more defensive when trading stocks that have recently IPO'd given their track record. Maybe I could use an RSI indicator to flag when an IPO stock is oversold so that I would know that it's been discounted from the initial price. Even then, I would further narrow down the stocks to does that I know has an established moat and I've done my research on.

In [108]:
# exploratory analysis, find 2025 IPO stocks that have been profitable after 12 months and have grown in their 12 month compared to their 1 month
pos_growth_12m = merged_df[
    (merged_df["future_growth_12_m"] > merged_df["future_growth_1_m"]) &
     (merged_df["future_growth_12_m"] > 1.0)
]
pos_growth_12m.head()

,ticker_x,min_date,Close,close_1st_day,future_growth_1_m,future_growth_2_m,future_growth_3_m,future_growth_4_m,future_growth_5_m,future_growth_6_m,future_growth_7_m,future_growth_8_m,future_growth_9_m,future_growth_10_m,future_growth_11_m,future_growth_12_m
4,CEPF,2025-08-21,10.12,10.12,1.004941,1.038538,1.023715,1.017787,1.032609,1.031621,1.010870,1.013834,1.014822,1.023715,1.037549,1.026186
9,BUUU,2025-08-14,4.00,4.00,1.825000,2.100000,2.185000,2.000000,2.097500,3.406250,3.855000,3.928750,5.250000,4.481250,5.477500,6.887500
10,MIAX,2025-08-14,30.74,30.74,1.212101,1.417046,1.657775,1.388419,1.391021,1.370852,1.295381,1.459011,1.826936,1.309044,1.377033,1.379636
11,RYOJ,2025-08-14,2.90,2.90,0.675862,0.662069,1.506896,1.375862,1.137931,0.958621,0.796552,0.697241,0.700000,0.806896,1.448276,1.572414
14,MAGH,2025-08-12,4.06,4.06,0.298768,0.315271,0.342365,1.665025,1.665025,1.665025,1.665025,1.665025,1.665025,1.665025,1.665025,1.665025


In [98]:
pos_growth_12m.describe()

,min_date,Close,close_1st_day,future_growth_1_m,future_growth_2_m,future_growth_3_m,future_growth_4_m,future_growth_5_m,future_growth_6_m,future_growth_7_m,future_growth_8_m,future_growth_9_m,future_growth_10_m,future_growth_11_m,future_growth_12_m
count,32,32.000000,32.000000,32.000000,32.000000,32.000000,32.000000,32.000000,32.000000,32.000000,32.000000,32.000000,32.000000,32.000000,32.000000
mean,2025-05-24 01:30:00,14.548223,14.671250,1.005651,1.141845,1.332085,1.393424,1.587552,1.785197,1.762878,1.716295,2.590755,2.892379,2.659867,2.611284
min,2025-01-07 00:00:00,2.900000,2.900000,0.298768,0.315271,0.342365,0.476119,0.597015,0.577335,0.710448,0.659486,0.617910,0.671642,0.728358,1.026186
25%,2025-03-02 06:00:00,4.197500,4.197500,0.912078,0.825415,0.844628,0.988517,0.966814,0.957994,0.971904,1.018325,1.033920,1.094458,1.171931,1.224195
50%,2025-06-26 00:00:00,12.585000,12.585000,1.005429,1.031421,1.071397,1.211590,1.217034,1.261517,1.213909,1.252441,1.349255,1.403747,1.509138,1.479713
75%,2025-08-02 12:00:00,19.100124,20.272500,1.114392,1.139775,1.517896,1.582739,1.717155,1.994119,2.010073,1.937098,2.046038,2.374570,2.639530,2.608902
max,2025-08-21 00:00:00,40.000000,40.000000,1.825000,3.077869,4.076500,3.417073,6.928943,8.760832,7.381283,5.232236,27.627298,34.687666,19.595802,22.044619
std,NaN,11.326535,11.373941,0.266145,0.591315,0.792948,0.707417,1.211785,1.521415,1.354993,1.120773,4.718336,5.910682,3.386417,3.760918


In [107]:
filtered_df[filtered_df.Symbol.isin(pos_growth_12m.ticker_x)]

,Company,Symbol,Industry,Offer Date,Shares (millions),Offer Price,1st Day Close,Current Price,Return,SCOOP Rating,close_1st_day
72,"Cantor Equity Partners IV, Inc. (Stock-Only SPAC)",CEPF,Blank Check,2025-08-21,40.0,$10.00,$10.12,$10.12,1.20,S/O,10.12
77,BUUU Group Limited,BUUU,Other,2025-08-14,1.5,$4.00,$4.00,$9.59,139.75,S/O,4.00
78,"Miami International Holdings, Inc. (Priced)",MIAX,Financials,2025-08-14,15.0,$23.00,$30.74,$42.23,83.61,S/O,30.74
79,rYojbaba Co. Ltd.,RYOJ,Health Care,2025-08-14,1.3,$4.00,$2.90,$2.90,-27.50,S/O,2.90
82,Magnitude International Ltd.,MAGH,Other,2025-08-12,2.2,$4.00,$4.06,$4.06,1.50,S/O,4.06
83,Avidbank Holdings (Uplisting – NASDAQ),AVBH,Financials,2025-08-08,2.6,$23.00,$23.25,$23.25,1.09,S/O,23.25
85,Etoiles Capital Group,EFTY,Other,2025-08-08,1.4,$4.00,$4.88,$4.88,22.00,S/O,4.88
88,"WhiteFiber, Inc.",WYFI,Technology,2025-08-07,9.4,$17.00,$16.22,$15.30,-10.00,S/O,16.22
90,HCM III Acquisition Corp.,HCMAU,Blank Check,2025-08-01,22.0,$10.00,$10.09,$10.10,1.00,S/O,10.09
92,"Shoulder Innovations, Inc.",SI,Health Care,2025-07-31,5.0,$15.00,$15.05,$14.39,-4.07,S/O,15.05
